src/segment-loader.js

```diff
@@ -819,32 +823,43 @@ export default class SegmentLoader extends videojs.EventTarget {
    */
   updateTimeline_(segmentInfo) {
     let segment;
-    let timelineUpdate;
+    let segmentEnd;
+    let timelineUpdated;
+    let segmentLength = this.playlist_.targetDuration;
     let playlist = segmentInfo.playlist;
     let currentMediaIndex = segmentInfo.mediaIndex;
 
     currentMediaIndex += playlist.mediaSequence - this.playlist_.mediaSequence;
     segment = playlist.segments[currentMediaIndex];
 
-    if (!segment) {
-      return;
-    }
-
-    timelineUpdate = Ranges.findSoleUncommonTimeRangesEnd(segmentInfo.buffered,
-                                                          this.sourceUpdater_.buffered());
-
     // Update segment meta-data (duration and end-point) based on timeline
-    let timelineUpdated = updateSegmentMetadata(playlist,
-                                                currentMediaIndex,
-                                                timelineUpdate);
+    if (segment &&
+        segmentInfo &&
+        segmentInfo.playlist.uri === this.playlist_.uri) {
+      segmentEnd = Ranges.findSoleUncommonTimeRangesEnd(segmentInfo.buffered,
+                                                        this.sourceUpdater_.buffered());
+      timelineUpdated = updateSegmentMetadata(playlist,
+                                              currentMediaIndex,
+                                              segmentEnd);
+      segmentLength = segment.duration;
+    }
 
     // the last segment append must have been entirely in the
     // already buffered time ranges. adjust the timeCorrection
     // offset to fetch forward until we find a segment that adds
     // to the buffered time ranges and improves subsequent media
     // index calculations.
     if (!timelineUpdated) {
-      this.timeCorrection_ -= segment.duration;
+      // appends haven't produced any new information for at least 5
+      // consecutive segments loads it is time to signal an error
+      // and stop
+      if (this.timeCorrection_ > this.playlist_.targetDuration * 5) {
+        this.timeCorrection_ = 0;
+        this.pause();
+        return this.trigger('error');
```

Comment:
We should definitely add tests that verify the error is triggered and loading paused when this condition is reached.